DEV-GUIDELINES.md

```diff
@@ -0,0 +1,187 @@
+        1. For pom packaging, the module is a "project module"
+    2. Maven coordinates:
+        1. Group id: io.helidon.${project_module}* - such as io.helidon.webserver; io.helidon.microprofile.config
+        2. Artifact id: helidon-${module_name}(-project)? - such as "helidon-security", "helicon-security-project", 
+            where project modules use the suffix "-project"
+        3. Version: always inherited
+    3. Package names:
+        1. io.helidon.${project_module}*.${module_name} - e.g. io.helidon.security, io.helidon.security.providers.common
+
+# Configuration and programmatic API
+1. **Everything that can be done using config, must be possible using programmatic approach through builders**
+    1. Everything that can be done using builders should be possible also using configuration, 
+        except for cases that would mandate usage of reflection 
+        (such an exception may be configuration of Routing in WebServer - nevertheless we still may support 
+            it (emphasis on "may" rather than "should"), or configuration of security for Jersey resources)
+2. When accepting config as a parameter, we should expect the config is located on the node that contains our configuration
+    (such as in ServerConfiguration in WebServer)
+3. Config keys:
+    1. Use lower case words separated by dashes 
+        (e.g. "token-endpoint-uri", NOT "tokenEndpointUri")
+    2. May be nested in a tree structure (e.g. outbound-token.name, outbound-token.algorithm)
+    3. Each component should have useful defaults defined for each property it expects. Components should only fail when missing 
+         a configuration that cannot be defaulted
+
+Example: [io.helidon.security.oidc.common.OidcConfig](security/providers/oidc-common/src/main/java/io/helidon/security/oidc/common/OidcConfig.java)
+
+# Getters and Setters
+1. We do not use the verb, e.g. when a property "port" exists, the following methods are used:  
+    1. port(int newPort)
+    2. int port()
+2. Boolean depends on how well understood the method is
+    1. Default is without a verb (e.g. authenticate(boolean atn), boolean authenticate())
+    2. If this would be ambiguous, we can use verb to clear the meaning (e.g. isAuthenticated() or shouldAuthenticate())      
+
+Example: [io.helidon.security.oidc.common.OidcConfig](security/providers/oidc-common/src/main/java/io/helidon/security/oidc/common/OidcConfig.java) 
+
+# Fluent API
+1. We use fluent API where applicable
+    1. In builders (see builders section below)
+    2. When using control methods (such as Server server = Server.create().start())
+    
+# Builders
+1. We use builders to create instances that need any parameters for construction
+    1. **This implies that there are no public API classes that would use public constructors**
+    2. Allowed exceptions to this rule:
+        1. Integration APIs that follow rules of integrated solution, e.g. Jersey SecurityFeature
```

Comment:
no, it was meant for using a different approach because of the integrated solution (e.g. Jersey commonly uses public default constructor in Features).
I am adding the exception you mentioned as well.